syntax = "proto3";

package network_analyzer;

service IngestService {
  // Streams packets from client to server
  rpc StreamPackets (stream Packet) returns (IngestSummary) {}
}

message Packet {
  double timestamp = 1;
  string interface = 2;
  string src_ip = 3;
  string dst_ip = 4;
  string protocol = 5;
  int32 length = 6;
  int32 src_port = 7;
  int32 dst_port = 8;
  string tcp_flags = 9;
  
  // Optional fields (using proto3 optional or default values)
  bool tcp_syn = 10;
  bool tcp_ack = 11;
  bool tcp_fin = 12;
  bool tcp_rst = 13;
  bool tcp_psh = 14;
  
  int64 seq = 15;
  int64 ack_num = 16;
  
  // ICMP
  int32 icmp_type = 17;
  int32 icmp_code = 18;
  
  // ARP
  int32 arp_op = 19;
  string arp_psrc = 20;
  string arp_pdst = 21;
  string arp_hwsrc = 22;
  string arp_hwdst = 23;
  
  // DNS
  bool dns_query = 24;
  string dns_qname = 25;
  int32 dns_qtype = 26;
  bool dns_response = 27;
  int32 dns_answer_count = 28;
  int32 dns_answer_size = 29;
  
  // HTTP
  string http_method = 30;
  string http_path = 31;
  string http_status_code = 32;
  string http_host = 33;
}

message IngestSummary {
  int32 received_count = 1;
  string status = 2;
  double server_timestamp = 3;
}
